shared/pubsub.js

```diff
@@ -412,9 +417,22 @@ const defaultMessageHandlers = {
     }, client.options.pingTimeout)
   },
 
-  [NOTIFICATION_TYPE.PUB] ({ channelID, data }) {
-    console.log(`[pubsub] Received data from channel ${channelID}:`, data)
-    // No need to reply.
+  [NOTIFICATION_TYPE.PUB] (msg) {
+    const { channelID, data } = msg
+
+    switch (data.type) {
+      case PUBSUB_EVENT_TYPE.CHATROOM_USER_TYPING: {
+        sbp('okTurtles.events/emit', PUBSUB_EVENT_TYPE.CHATROOM_USER_TYPING, { username: data.username })
+        break
+      }
+      case PUBSUB_EVENT_TYPE.CHATROOM_USER_STOP_TYPING: {
+        sbp('okTurtles.events/emit', PUBSUB_EVENT_TYPE.CHATROOM_USER_STOP_TYPING, { username: data.username })
+        break
+      }
+      default: {
+        console.log(`[pubsub] Received data from channel ${channelID}:`, data)
+      }
+    }
```

Comment:
Excellent, this PR is starting to shape up nicely.

There is one important thing to keep in mind though: these handlers need to be part of Group Income, and this code in `shared/pubsub.js` is going to be part of Chelonia. The two must be kept separate to make it easier to split out Chelonia in the future.

This is a great opportunity to improve the framework though. We made an oversight: we forgot to handle setting custom `messageHandlers` in `'chelonia/connect'`.

Currently, the code in there looks contains this:

```js
    this.pubsub = createClient(pubsubURL, {
      ...this.config.connectionOptions,
      messageHandlers: {
        [NOTIFICATION_TYPE.ENTRY] (msg) {
          // We MUST use 'chelonia/private/in/enqueueHandleEvent' to ensure handleEvent()
          // is called AFTER any currently-running calls to 'chelonia/contract/sync'
          // to prevent gi.db from throwing "bad previousHEAD" errors.
          // Calling via SBP also makes it simple to implement 'test/backend.js'
          const { contractID } = GIMessage.deserializeHEAD(msg.data)
          sbp('chelonia/private/in/enqueueHandleEvent', contractID, msg.data)
        },
        [NOTIFICATION_TYPE.VERSION_INFO] (msg) {
          const ourVersion = process.env.GI_VERSION
          const theirVersion = msg.data.GI_VERSION

          const ourContractsVersion = process.env.CONTRACTS_VERSION
          const theirContractsVersion = msg.data.CONTRACTS_VERSION
          if (ourVersion !== theirVersion || ourContractsVersion !== theirContractsVersion) {
            sbp('okTurtles.events/emit', NOTIFICATION_TYPE.VERSION_INFO, { ...msg.data })
          }
        },
        [REQUEST_TYPE.PUSH_ACTION] (msg) {
          sbp('okTurtles.events/emit', REQUEST_TYPE.PUSH_ACTION, { data: msg.data })
        }
      }
    })
```

From here we need to do several things here to prepare Chelonia for release:

1. Move `NOTIFICATION_TYPE.VERSION_INFO` and `REQUEST_TYPE.PUSH_ACTION` handlers out of there, and instead include them as part of an object that gets merged into `messageHandlers`, and pass that object in via the `'chelonia/connect'` parameter.
2. Undo your changes here to `NOTIFICATION_TYPE.PUB`, and move that code as part of the parameter that gets passed in to `'chelonia/connect'` (with the other 2 handlers)

So in the end you'll have something that looks something like this:

```js
  'chelonia/connect': function ({ messageHandlers } = {}): Object {
```

And then that gets merged below:

```js
    this.pubsub = createClient(pubsubURL, {
      ...this.config.connectionOptions,
      messageHandlers: {
        ...messageHandlers,
        [NOTIFICATION_TYPE.ENTRY] (msg) {
          // We MUST use 'chelonia/private/in/enqueueHandleEvent' to ensure handleEvent()
          // is called AFTER any currently-running calls to 'chelonia/contract/sync'
          // to prevent gi.db from throwing "bad previousHEAD" errors.
          // Calling via SBP also makes it simple to implement 'test/backend.js'
          const { contractID } = GIMessage.deserializeHEAD(msg.data)
          sbp('chelonia/private/in/enqueueHandleEvent', contractID, msg.data)
        }
```

And then update `main.js` with the handlers.